File: protocol/impls/csharp/AttribCurve.cs
using System;
using System.Collections.Generic;
using System.Security.Cryptography;
using System.Text;

namespace Shibpost;

/// <summary>
/// §4 Strategy B — the pinned ECDSA curve-vector set (`sm attrib-curve`).
/// Mirrors impls/c/src/attrib_curve.c (attrib_cmd_curve) + the e2e in attrib.c
/// (attrib_real_endtoend). Prints byte-identical output to the C reference:
/// pinned P/N/N_HALF, on-curve edges, RFC-6979 signers + verify-boundary battery,
/// tiny-key KAT, PRIMARY `combined` digest, then 3 end-to-end txs + `combined_e2e`.
/// </summary>
public static class AttribCurve
{
    private static readonly byte[] CV_P     = Hex("FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F");
    private static readonly byte[] CV_N     = Hex("FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141");
    private static readonly byte[] CV_NHALF = Hex("7FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF5D576E7357A4501DDFE92F46681B20A0");
    private static readonly byte[] CV_GX    = Hex("79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798");
    private static readonly byte[] CV_GY    = Hex("483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A68554199C47D08FFB10D4B8");

    public static int Run()
    {
        var sb = new StringBuilder();
        var comb = IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
        void Feed(byte[] d, int n) { comb.AppendData(d, 0, n); }
        void Feed1(byte b) => comb.AppendData(new[] { b }, 0, 1);

        // ── 1. pinned constants ──────────────────────────────────────────────
        sb.Append("p ").Append(Hx(CV_P)).Append('\n');         Feed(CV_P, 32);
        sb.Append("n ").Append(Hx(CV_N)).Append('\n');         Feed(CV_N, 32);
        sb.Append("nhalf ").Append(Hx(CV_NHALF)).Append('\n'); Feed(CV_NHALF, 32);

        // ── 2. on-curve membership at the edges ──────────────────────────────
        var oc = new List<(string name, byte[] key)>();
        { byte[] b = new byte[65]; b[0] = 0x04; Copy(b, 1, CV_GX); Copy(b, 33, CV_GY); oc.Add(("oc_G_uncomp", b)); }
        { byte[] b = new byte[33]; b[0] = 0x02; Copy(b, 1, CV_GX); oc.Add(("oc_G_comp02", b)); }
        { byte[] b = new byte[33]; b[0] = 0x03; Copy(b, 1, CV_GX); oc.Add(("oc_G_comp03", b)); }
        { byte[] b = new byte[65]; b[0] = 0x04; Copy(b, 1, CV_GX); Copy(b, 33, CV_GY); b[64] ^= 0x01; oc.Add(("oc_G_badY", b)); }
        { byte[] b = new byte[33]; b[0] = 0x02; oc.Add(("oc_X0", b)); }                       // X=0
        { byte[] b = new byte[33]; b[0] = 0x02; b[32] = 1; oc.Add(("oc_X1", b)); }            // X=1
        { byte[] b = new byte[65]; b[0] = 0x04; Copy(b, 1, CV_P); Copy(b, 33, CV_GY); oc.Add(("oc_Xeqp", b)); }
        { byte[] b = new byte[33]; b[0] = 0x02; Copy(b, 1, CV_P); oc.Add(("oc_comp_Xeqp", b)); }
        { byte[] b = new byte[33]; b[0] = 0x05; Copy(b, 1, CV_GX); oc.Add(("oc_badprefix", b)); }
        foreach (var (name, key) in oc)
        {
            int v = Secp256k1.OnCurve(key) ? 1 : 0;
            sb.Append(name).Append(' ').Append(v).Append('\n');
            Feed1((byte)v); Feed(key, key.Length);
        }

        // ── 3 & 4. RFC-6979 deterministic sign + ECDSA verify at the boundaries ─
        for (int i = 0; i < 4; i++)
        {
            byte[] priv = new byte[32];
            priv[28] = 0xC0; priv[29] = 0xFF; priv[30] = 0xEE; priv[31] = (byte)(0x10 + i);
            if (!Secp256k1.Pubkey(priv, out byte[] pub)) { sb.Append("sig").Append(i).Append(" PUBFAIL\n"); continue; }
            byte[] m = Encoding.ASCII.GetBytes("strategy-b curve vector " + i);
            byte[] h = Hashing.Sha256(m);
            if (!Secp256k1.EcdsaSign(priv, h, out byte[] r, out byte[] s)) { sb.Append("sig").Append(i).Append(" SIGNFAIL\n"); continue; }
            byte[] der = DerSig(r, s);
            sb.Append("sig").Append(i).Append(" pub=").Append(Hx(pub))
              .Append(" r=").Append(Hx(r))
              .Append(" s=").Append(Hx(s))
              .Append(" der=").Append(Hx(der)).Append('\n');
            Feed(pub, 33); Feed(r, 32); Feed(s, 32); Feed(der, der.Length);

            // verify boundary battery
            byte[] zero = new byte[32];
            byte[] hbad = (byte[])h.Clone(); hbad[0] ^= 0x01;
            byte[] hiS = new byte[32];                          // high-S = n - s (byte subtraction)
            { int borrow = 0; for (int k = 31; k >= 0; k--) { int d = CV_N[k] - s[k] - borrow; if (d < 0) { d += 256; borrow = 1; } else borrow = 0; hiS[k] = (byte)d; } }
            byte[] wrongpub = (byte[])pub.Clone(); wrongpub[0] ^= 0x01;
            var vt = new (string nm, byte[] hh, byte[] rr, byte[] ss, byte[] pk)[]
            {
                ("valid",   h,    r,    s,    pub),
                ("tamper",  hbad, r,    s,    pub),
                ("r0",      h,    zero, s,    pub),
                ("s0",      h,    r,    zero, pub),
                ("rN",      h,    CV_N, s,    pub),
                ("sN",      h,    r,    CV_N, pub),
                ("highS",   h,    r,    hiS,  pub),
                ("wrongpk", h,    r,    s,    wrongpub),
            };
            sb.Append("ver").Append(i);
            foreach (var t in vt)
            {
                int v = Secp256k1.EcdsaVerify(t.hh, t.rr, t.ss, t.pk) ? 1 : 0;
                sb.Append(' ').Append(t.nm).Append('=').Append(v);
                Feed1((byte)v);
            }
            sb.Append('\n');
        }

        // ── 5. tiny-key KAT: priv=1 ⇒ pub=G ; priv=2 ⇒ pub=2G ────────────────
        { byte[] p1 = new byte[32]; p1[31] = 1; Secp256k1.Pubkey(p1, out byte[] pk1); sb.Append("priv1_pub=").Append(Hx(pk1)).Append('\n'); Feed(pk1, 33); }
        { byte[] p2 = new byte[32]; p2[31] = 2; Secp256k1.Pubkey(p2, out byte[] pk2); sb.Append("priv2_pub=").Append(Hx(pk2)).Append('\n'); Feed(pk2, 33); }

        byte[] cd = comb.GetHashAndReset();
        sb.Append("combined ").Append(Hx(cd)).Append('\n');

        // ── 6. end-to-end (attrib_real_endtoend) ─────────────────────────────
        var e2e = IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
        EndToEnd(sb, e2e);
        byte[] ed = e2e.GetHashAndReset();
        sb.Append("combined_e2e ").Append(Hx(ed)).Append('\n');

        Console.Out.Write(sb.ToString());
        return 0;
    }

    // mirror C der_int/der_sig (canonical strict-DER ‖ SIGHASH_ALL).
    private static byte[] DerInt(byte[] v)
    {
        int i = 0; while (i < 31 && v[i] == 0) i++;
        int len = 32 - i; int pad = (v[i] & 0x80) != 0 ? 1 : 0;
        byte[] outp = new byte[2 + pad + len];
        int n = 0;
        outp[n++] = 0x02; outp[n++] = (byte)(len + pad);
        if (pad == 1) outp[n++] = 0x00;
        Buffer.BlockCopy(v, i, outp, n, len);
        return outp;
    }
    private static byte[] DerSig(byte[] r, byte[] s)
    {
        byte[] ri = DerInt(r), si = DerInt(s);
        int bl = ri.Length + si.Length;
        byte[] outp = new byte[2 + bl + 1];
        outp[0] = 0x30; outp[1] = (byte)bl;
        Buffer.BlockCopy(ri, 0, outp, 2, ri.Length);
        Buffer.BlockCopy(si, 0, outp, 2 + ri.Length, si.Length);
        outp[2 + bl] = 0x01; // SIGHASH_ALL
        return outp;
    }

    // ── end-to-end: sign the actual legacy sighash, attribute() with real curve ──
    private static void EndToEnd(StringBuilder sb, IncrementalHash comb)
    {
        bool prev = Attribution.RealCurve;
        Attribution.RealCurve = true;
        try
        {
            void Emit(string name, RawTx tx)
            {
                AttribResult res = Attribution.Attribute(tx, 0);
                sb.Append(name).Append(' ').Append(res.Status).Append(':').Append(Hx(res.Identity)).Append('\n');
                comb.AppendData(new[] { (byte)res.Status }, 0, 1);
                comb.AppendData(res.Sighash, 0, 32);
                comb.AppendData(res.Identity, 0, 20);
            }

            // A. P2PKH, correctly signed ⇒ FOUND (priv = 0x2A)
            {
                byte[] priv = new byte[32]; priv[31] = 0x2A;
                Secp256k1.Pubkey(priv, out byte[] pub);
                byte[] h160 = Hashing.Hash160(pub);
                byte[] sc = Attribution.P2pkhScriptCode(h160);
                byte[] sh = SkeletonSighash(sc);
                Secp256k1.EcdsaSign(priv, sh, out byte[] r, out byte[] s);
                byte[] der = DerSig(r, s);
                byte[] ss = Concat(EmitPush(der), EmitPush(pub));
                Emit("e2e_p2pkh_valid", SkeletonTx(ss));
            }
            // B. P2PKH, signed by the WRONG key ⇒ verify-drop (status 2)
            {
                byte[] priv = new byte[32]; priv[31] = 0x2A;
                byte[] wrong = new byte[32]; wrong[31] = 0x2B;
                Secp256k1.Pubkey(priv, out byte[] pub);
                byte[] h160 = Hashing.Hash160(pub);
                byte[] sc = Attribution.P2pkhScriptCode(h160);
                byte[] sh = SkeletonSighash(sc);
                Secp256k1.EcdsaSign(wrong, sh, out byte[] r, out byte[] s);
                byte[] der = DerSig(r, s);
                byte[] ss = Concat(EmitPush(der), EmitPush(pub));
                Emit("e2e_p2pkh_wrongkey", SkeletonTx(ss));
            }
            // C. 2-of-2 P2SH multisig, two correct in-order sigs ⇒ FOUND (privs 0x50,0x51)
            {
                byte[][] priv = new byte[2][]; byte[][] pub = new byte[2][];
                for (int i = 0; i < 2; i++) { priv[i] = new byte[32]; priv[i][31] = (byte)(0x50 + i); Secp256k1.Pubkey(priv[i], out pub[i]); }
                // redeem: OP_2 <k0><k1> OP_2 OP_CHECKMULTISIG
                var rs = new List<byte> { 0x52 };
                for (int i = 0; i < 2; i++) { rs.Add(0x21); rs.AddRange(pub[i]); }
                rs.Add(0x52); rs.Add(0xAE);
                byte[] redeem = rs.ToArray();
                byte[] sh = SkeletonSighash(redeem);
                var ss = new List<byte> { 0x00 };                // NULLDUMMY
                for (int i = 0; i < 2; i++)
                {
                    Secp256k1.EcdsaSign(priv[i], sh, out byte[] r, out byte[] s);
                    ss.AddRange(EmitPush(DerSig(r, s)));
                }
                ss.AddRange(EmitPush(redeem));
                Emit("e2e_multisig_valid", SkeletonTx(ss.ToArray()));
            }
        }
        finally { Attribution.RealCurve = prev; }
    }

    // shared tx skeleton: 1 input (outpoint 0x11.., seq FFFFFFFF), 1 output (value 100000,
    // scriptPubKey = OP_RETURN), version 1, locktime 0.
    private static RawTx SkeletonTx(byte[] scriptSig)
    {
        byte[] prev = new byte[32]; for (int i = 0; i < 32; i++) prev[i] = 0x11;
        return new RawTx
        {
            Version = 1,
            Inputs = { new RawInput { PrevHash = prev, PrevN = 0x11111111, ScriptSig = scriptSig, Sequence = 0xFFFFFFFF } },
            Outputs = { new RawOutput { Value = 100000, ScriptPubKey = new byte[] { 0x6a } } },
            Locktime = 0,
        };
    }
    private static byte[] SkeletonSighash(byte[] scriptCode)
    {
        // the legacy sighash computed by attribute() over the same skeleton (no scriptSig sigs to FaD here).
        return Attribution.LegacySighash(SkeletonTx(Array.Empty<byte>()), 0, scriptCode, Array.Empty<byte[]>());
    }

    private static byte[] EmitPush(byte[] data)
    {
        if (data.Length < 0x4C)
        {
            byte[] r = new byte[1 + data.Length]; r[0] = (byte)data.Length; Buffer.BlockCopy(data, 0, r, 1, data.Length); return r;
        }
        byte[] r2 = new byte[2 + data.Length]; r2[0] = 0x4C; r2[1] = (byte)data.Length; Buffer.BlockCopy(data, 0, r2, 2, data.Length); return r2;
    }

    // ── small helpers ────────────────────────────────────────────────────────
    private static void Copy(byte[] dst, int off, byte[] src) => Buffer.BlockCopy(src, 0, dst, off, src.Length);
    private static byte[] Concat(params byte[][] parts)
    {
        int len = 0; foreach (var p in parts) len += p.Length;
        byte[] r = new byte[len]; int o = 0; foreach (var p in parts) { Buffer.BlockCopy(p, 0, r, o, p.Length); o += p.Length; } return r;
    }
    private static byte[] Hex(string s) { byte[] r = new byte[s.Length / 2]; for (int i = 0; i < r.Length; i++) r[i] = Convert.ToByte(s.Substring(i * 2, 2), 16); return r; }
    private static string Hx(byte[] d) { var sb = new StringBuilder(d.Length * 2); foreach (var b in d) sb.Append(b.ToString("x2")); return sb.ToString(); }
}

File: protocol/impls/go/attrib_curve.go
package main

// §4 Strategy B — the pinned ECDSA curve-vector set (`sm attrib-curve`).
//
// Mirrors impls/c/src/attrib_curve.c + attrib_real_endtoend: pinned P/N/N_HALF
// constants, on-curve membership at the edges, RFC-6979 deterministic (r,s) +
// canonical-DER known-answers, ECDSA verify accept/reject at the scalar
// boundaries, a tiny-key KAT, the PRIMARY `combined` digest, and an end-to-end
// real-curve pipeline (`combined_e2e`). Output must be byte-identical across all
// 7 reference impls. See SPEC-conformance.md §13 + SPEC-RATIONALE.md §11.

import (
	"crypto/sha256"
	"fmt"
)

// curve-vector constants (mirror secp256k1.c; pinned so the script self-contains
// its expected scalars cross-language).
var (
	cvP    = mustHex32("FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F")
	cvN    = mustHex32("FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141")
	cvNH   = mustHex32("7FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF5D576E7357A4501DDFE92F46681B20A0")
	cvGx   = mustHex32("79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798")
	cvGy   = mustHex32("483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A68554199C47D08FFB10D4B8")
)

const hexd = "0123456789abcdef"

func hexstr(d []byte) string {
	out := make([]byte, 0, len(d)*2)
	for _, b := range d {
		out = append(out, hexd[b>>4], hexd[b&15])
	}
	return string(out)
}

// canonical strict-DER encoding of (r,s) ‖ SIGHASH_ALL.
func derIntCurve(v []byte) []byte {
	i := 0
	for i < 31 && v[i] == 0 {
		i++
	}
	body := v[i:]
	out := []byte{0x02}
	if body[0]&0x80 != 0 {
		out = append(out, byte(len(body)+1), 0x00)
	} else {
		out = append(out, byte(len(body)))
	}
	return append(out, body...)
}

func derSigCurve(r, s []byte) []byte {
	body := append(derIntCurve(r), derIntCurve(s)...)
	out := []byte{0x30, byte(len(body))}
	out = append(out, body...)
	return append(out, 0x01) // SIGHASH_ALL
}

func runAttribCurve() {
	comb := sha256.New()
	feed := func(b []byte) { comb.Write(b) }

	// ── 1. pinned constants ────────────────────────────────────────────────────
	fmt.Printf("p %s\n", hexstr(cvP[:]))
	feed(cvP[:])
	fmt.Printf("n %s\n", hexstr(cvN[:]))
	feed(cvN[:])
	fmt.Printf("nhalf %s\n", hexstr(cvNH[:]))
	feed(cvNH[:])

	// ── 2. on-curve membership at the edges ────────────────────────────────────
	type ocVec struct {
		name string
		key  []byte
	}
	var ocs []ocVec
	add := func(name string, key []byte) { ocs = append(ocs, ocVec{name, key}) }

	guncomp := append([]byte{0x04}, append(append([]byte{}, cvGx[:]...), cvGy[:]...)...)
	add("oc_G_uncomp", guncomp)
	add("oc_G_comp02", append([]byte{0x02}, cvGx[:]...))
	add("oc_G_comp03", append([]byte{0x03}, cvGx[:]...))
	badY := append([]byte{0x04}, append(append([]byte{}, cvGx[:]...), cvGy[:]...)...)
	badY[64] ^= 0x01
	add("oc_G_badY", badY)
	add("oc_X0", append([]byte{0x02}, make([]byte, 32)...))
	x1 := make([]byte, 32)
	x1[31] = 1
	add("oc_X1", append([]byte{0x02}, x1...))
	add("oc_Xeqp", append([]byte{0x04}, append(append([]byte{}, cvP[:]...), cvGy[:]...)...))
	add("oc_comp_Xeqp", append([]byte{0x02}, cvP[:]...))
	add("oc_badprefix", append([]byte{0x05}, cvGx[:]...))

	for _, o := range ocs {
		v := 0
		if secpOnCurve(o.key) {
			v = 1
		}
		fmt.Printf("%s %d\n", o.name, v)
		feed([]byte{byte(v)})
		feed(o.key)
	}

	// ── 3 & 4. RFC-6979 deterministic sign + ECDSA verify at the boundaries ─────
	for i := 0; i < 4; i++ {
		priv := make([]byte, 32)
		priv[28] = 0xC0
		priv[29] = 0xFF
		priv[30] = 0xEE
		priv[31] = byte(0x10 + i)
		pub, ok := secpPubkey(priv)
		if !ok {
			fmt.Printf("sig%d PUBFAIL\n", i)
			continue
		}
		m := fmt.Sprintf("strategy-b curve vector %d", i)
		h := sha256.Sum256([]byte(m))
		r32, s32, sok := secpEcdsaSign(priv, h[:])
		if !sok {
			fmt.Printf("sig%d SIGNFAIL\n", i)
			continue
		}
		der := derSigCurve(r32[:], s32[:])
		fmt.Printf("sig%d pub=%s r=%s s=%s der=%s\n", i, hexstr(pub), hexstr(r32[:]), hexstr(s32[:]), hexstr(der))
		feed(pub)
		feed(r32[:])
		feed(s32[:])
		feed(der)

		// verify boundary battery.
		var zero [32]byte
		var hh [32]byte
		copy(hh[:], h[:])
		var hbad [32]byte
		copy(hbad[:], h[:])
		hbad[0] ^= 0x01
		var hiS [32]byte // high-S = n - s
		borrow := 0
		for k := 31; k >= 0; k-- {
			d := int(cvN[k]) - int(s32[k]) - borrow
			if d < 0 {
				d += 256
				borrow = 1
			} else {
				borrow = 0
			}
			hiS[k] = byte(d)
		}
		wrongpub := append([]byte{}, pub...)
		wrongpub[0] ^= 0x01

		type vt struct {
			nm string
			h  [32]byte
			r  [32]byte
			s  [32]byte
			pk []byte
		}
		vts := []vt{
			{"valid", hh, r32, s32, pub},
			{"tamper", hbad, r32, s32, pub},
			{"r0", hh, zero, s32, pub},
			{"s0", hh, r32, zero, pub},
			{"rN", hh, cvN, s32, pub},
			{"sN", hh, r32, cvN, pub},
			{"highS", hh, r32, hiS, pub},
			{"wrongpk", hh, r32, s32, wrongpub},
		}
		fmt.Printf("ver%d", i)
		for _, t := range vts {
			v := 0
			if secpEcdsaVerify(t.h, t.r, t.s, t.pk) {
				v = 1
			}
			fmt.Printf(" %s=%d", t.nm, v)
			feed([]byte{byte(v)})
		}
		fmt.Printf("\n")
	}

	// ── 5. tiny-key KAT: priv=1 ⇒ pub=G ; priv=2 ⇒ pub=2G ──────────────────────
	p1 := make([]byte, 32)
	p1[31] = 1
	pk1, _ := secpPubkey(p1)
	fmt.Printf("priv1_pub=%s\n", hexstr(pk1))
	feed(pk1)
	p2 := make([]byte, 32)
	p2[31] = 2
	pk2, _ := secpPubkey(p2)
	fmt.Printf("priv2_pub=%s\n", hexstr(pk2))
	feed(pk2)

	// PRIMARY cross-language digest.
	cd := comb.Sum(nil)
	fmt.Printf("combined %s\n", hexstr(cd))

	// ── 6. end-to-end: sign the real legacy sighash, attribute() with real curve ──
	e2e := sha256.New()
	attribRealEndToEnd(e2e)
	ed := e2e.Sum(nil)
	fmt.Printf("combined_e2e %s\n", hexstr(ed))
}

// attribRealEndToEnd mirrors C attrib_real_endtoend: builds 3 txs (P2PKH valid,
// P2PKH wrong-key, 2-of-2 multisig), signs the genuine legacy sighash with
// RFC-6979, and runs the full attribute() pipeline with the REAL curve.
func attribRealEndToEnd(comb interface{ Write([]byte) (int, error) }) {
	gRealCurve = true
	defer func() { gRealCurve = false }()

	emit := func(name string, t rawTx, k int) {
		res := attribute(t, k)
		fmt.Printf("%s %d:%s\n", name, res.status, hexstr(res.identity[:]))
		comb.Write([]byte{byte(res.status)})
		comb.Write(res.sighash[:])
		comb.Write(res.identity[:])
	}

	// shared skeleton: input outpoint 0x11×36, seq FFFFFFFF, output value 100000
	// scriptPubKey OP_RETURN (0x6a), version 1, locktime 0.
	skeleton := func() rawTx {
		var t rawTx
		t.version = 1
		var op [32]byte
		for i := range op {
			op[i] = 0x11
		}
		t.prevTxid = append(t.prevTxid, op)
		t.prevVout = append(t.prevVout, 0x11111111)
		t.vin = append(t.vin, txIn{})
		t.seq = append(t.seq, 0xFFFFFFFF)
		t.vout = append(t.vout, txOut{value: 100000, script: []byte{0x6a}})
		return t
	}
	sighashOf := func(scriptCode []byte) [32]byte { return legacySighash(skeleton(), 0, scriptCode) }

	// assemble raw tx carrying scriptSig ss; same structural fields as skeleton.
	rawWith := func(ss []byte) []byte {
		var b []byte
		b = appendU32(b, 1)
		b = appendVarint(b, 1)
		for i := 0; i < 36; i++ {
			b = append(b, 0x11)
		}
		b = appendVarint(b, uint64(len(ss)))
		b = append(b, ss...)
		b = appendU32(b, 0xFFFFFFFF)
		b = appendVarint(b, 1)
		b = appendU64(b, 100000)
		b = appendVarint(b, 1)
		b = append(b, 0x6a)
		b = appendU32(b, 0)
		return b
	}

	// ── A. P2PKH, correctly signed ⇒ FOUND (3) ──────────────────────────────────
	{
		priv := make([]byte, 32)
		priv[31] = 0x2A
		pub, _ := secpPubkey(priv)
		h160 := hash160(pub)
		sc := p2pkhScript(h160)
		sh := sighashOf(sc)
		r32, s32, _ := secpEcdsaSign(priv, sh[:])
		der := derSigCurve(r32[:], s32[:])
		ss := append(pushEncode(der), pushEncode(pub)...)
		if t, ok := parseTx(rawWith(ss)); ok {
			emit("e2e_p2pkh_valid", t, 0)
		} else {
			fmt.Println("e2e_p2pkh_valid PARSEFAIL")
		}
	}
	// ── B. P2PKH, signed by the WRONG key ⇒ verify-drop (2) ─────────────────────
	{
		priv := make([]byte, 32)
		priv[31] = 0x2A
		wrong := make([]byte, 32)
		wrong[31] = 0x2B
		pub, _ := secpPubkey(priv)
		h160 := hash160(pub)
		sc := p2pkhScript(h160)
		sh := sighashOf(sc)
		r32, s32, _ := secpEcdsaSign(wrong, sh[:])
		der := derSigCurve(r32[:], s32[:])
		ss := append(pushEncode(der), pushEncode(pub)...)
		if t, ok := parseTx(rawWith(ss)); ok {
			emit("e2e_p2pkh_wrongkey", t, 0)
		} else {
			fmt.Println("e2e_p2pkh_wrongkey PARSEFAIL")
		}
	}
	// ── C. 2-of-2 P2SH multisig, two in-order sigs ⇒ FOUND (3) ──────────────────
	{
		var pubs [2][]byte
		var privs [2][]byte
		for i := 0; i < 2; i++ {
			priv := make([]byte, 32)
			priv[31] = byte(0x50 + i)
			privs[i] = priv
			pubs[i], _ = secpPubkey(priv)
		}
		// OP_2 <k0><k1> OP_2 OP_CHECKMULTISIG
		rs := []byte{0x52}
		for i := 0; i < 2; i++ {
			rs = append(rs, 0x21)
			rs = append(rs, pubs[i]...)
		}
		rs = append(rs, 0x52, 0xae)
		sh := sighashOf(rs)
		ss := []byte{0x00} // NULLDUMMY
		for i := 0; i < 2; i++ {
			r32, s32, _ := secpEcdsaSign(privs[i], sh[:])
			ss = append(ss, pushEncode(derSigCurve(r32[:], s32[:]))...)
		}
		ss = append(ss, pushEncode(rs)...)
		if t, ok := parseTx(rawWith(ss)); ok {
			emit("e2e_multisig_valid", t, 0)
		} else {
			fmt.Println("e2e_multisig_valid PARSEFAIL")
		}
	}
}
